### PR TITLE
Fix 500 error upon GET request to /hello, closes #44

### LINKED ISSUE #44: 500 Error /hello endpoint [High]
**Bug Summary**

Getting 500 server error when hitting /hello (currently mapped to reddit flow).
See screenshot below.

**Test Case(s)**
N/A (we should have integration tests for this)

---

**Expected behavior**

"hello" string returned as JSON object.
Logger outputs reddit posts.

**Observed Behaviour**

500 server error

**Steps To Reproduce**

1. Populate application.properties with reddit credentials
2. Start spring server
3. Make a GET to /hello 

**Environment**
- Version: latest
- OS: macos (latest)
- Browser: chrome (latest)

---

**Code Examples**

**Stack Trace**
```
java.lang.NullPointerException: null
	at com.feeder.server.providers.reddit.RedditFeedProvider.getFeed(RedditFeedProvider.java:61) ~[classes/:na]
	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
Error has been observed at the following site(s):
	|_ checkpoint ⇢ HTTP GET "/hello" [ExceptionHandlingWebHandler]
Stack trace:
		at com.feeder.server.providers.reddit.RedditFeedProvider.getFeed(RedditFeedProvider.java:61) ~[classes/:na]
		at com.feeder.server.controllers.HelloController.readHello(HelloController.java:28) ~[classes/:na]
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
		at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
		at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
		at org.springframework.web.reactive.result.method.InvocableHandlerMethod.lambda$invoke$0(InvocableHandlerMethod.java:147) ~[spring-webflux-5.2.4.RELEASE.jar:5.2.4.RELEASE]
		at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:151) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
		at reactor.core.publisher.MonoFlatMap.subscribeOrReturn(MonoFlatMap.java:53) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
		at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:48) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
		at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
		at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.drain(MonoIgnoreThen.java:153) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
		at reactor.core.publisher.MonoIgnoreThen.subscribe(MonoIgnoreThen.java:56) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
		at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:55) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
		at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:150) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
		at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:67) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
		at reactor.core.publisher.MonoNext$NextSubscriber.onNext(MonoNext.java:76) ~[reactor-core-3.3.3.RELEASE.jar:3.3.3.RELEASE]
...
```
**Screenshots**
image

**Error Report**

---

**Additional context**

### PR DESCRIPTION
**Pull request by HarrisonLeach1**
GitHub Issue (If applicable): #44

## PR Type

What kind of change does this PR introduce?

- Bugfix
## What is the current behavior?
- 500 error upon hitting the `/hello` endpoint

## What is the new behavior?
- `/hello` endpoint returns `200 OK` response with Reddit API data collected

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current upstream repo
- [x] Added new tests to the test suite for the feature
- [x] Tested the full test suite, and ensured that the feature does not break current build.
- [ ] Docs have been added/updated which fit (for bug fixes / features)
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal)
- [x] Checked licensing of Frameworks/Libraries (if applicable) 

## Other information
- This was caused by RedditFeedProvider being instantiated via spring IoC
using the parameterless constructor. This constructor was only meant to be
used for unit tests. To fix this a Factory was made to better control the
complex creation of the RedditFeedProvider.

Internal Issue (If applicable):

### CODE REVIEW

File: server/src/main/java/com/feeder/server/controllers/HelloController.java [hunk 1]
```diff
@@ -19,12 +20,14 @@
 public class HelloController {
 
     private static final Logger logger = LoggerFactory.getLogger(HelloController.class);
-    @Autowired private ApplicationProperties properties;
-    @Autowired private RedditFeedProvider redditFeedProvider;
+    @Autowired private RedditFeedProviderFactory redditFeedProviderFactory;
+    @Autowired private ApplicationProperties applicationProperties;
```
Comment (jordansimsmith) [inline server/src/main/java/com/feeder/server/controllers/HelloController.java]: From what I've seen, I think its more standard to have the annotations above the field? (thats how intellij formats it atleast). However, not really an issue and definitely not blocking. Guidelines will become apparent when #13 is resolved.

File: server/src/test/java/com/feeder/server/providers/reddit/RedditFeedProviderFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    public void testPropertyValidation() {
+        try {
+            subject.createProvider(new ApplicationProperties("","","",""));
+            fail();
+        } catch (IllegalStateException e) {
+
+        }
+    }
+
+    public class TestableRedditFeedProviderFactory extends RedditFeedProviderFactory {
```
Comment (jordansimsmith) [inline server/src/test/java/com/feeder/server/providers/reddit/RedditFeedProviderFactoryTest.java]: could this class be private?

File: server/src/test/java/com/feeder/server/providers/reddit/RedditFeedProviderFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    private TestableRedditFeedProviderFactory subject;
+
+
+    @BeforeEach
+    void setUp() {
+        subject = new TestableRedditFeedProviderFactory();
+    }
+
+    @Test
+    void testCreateProvider() {
```
Comment (jordansimsmith) [inline server/src/test/java/com/feeder/server/providers/reddit/RedditFeedProviderFactoryTest.java]: Same as before with the visibility modifiers

**Comment by HarrisonLeach1** [comment]
This PR will be closed, as an alternative approach will be taken.

**Pull request closed without merging**